parlai/core/opt.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Opt is the system for passing around options throughout ParlAI.
+"""
+
+import copy
+import json
+import pickle
+import traceback
+
+
+class Opt(dict):
+    """
+    Class for tracking options.
+
+    Functions like a dict, but allows us to track the history of arguments as they are
+    set.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.history = {}
+        self.deepcopies = []
+
+    def __setitem__(self, key, val):
+        loc = traceback.format_stack()[-2]
+        self.history.setdefault(key, []).append((loc, val))
```

Comment:
This is the first time I've seen the Opt implementation - it looks super useful! I assume the second-to-last line of the stack is the call just before `__setitem__()`? (I'm trying to think if there'd be a need to have different opts for different calls of the same function, which would cause a collision here, but that's probably an edge case)

Reply:
-1 is the current line (opt.py:30), -2 is whatever called `opt[key] = value`.